walk-in-interview/src/main/java/com/google/job/servlets/JobsListingsServlet.java

```diff
@@ -0,0 +1,229 @@
+package com.google.job.servlets;
+
+import com.google.job.data.*;
+import com.google.gson.Gson;
+
+import javax.servlet.ServletException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import java.io.IOException;
+
+import java.sql.Time;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.*;
+
+/**
+ * Servlet that handles getting the job listings given the filters.
+ */
+@WebServlet("/jobs/listings")
+public final class JobsListingsServlet extends HttpServlet {
+    private static final long TIMEOUT = 5;
+
+    public static final String MIN_LIMIT_PARAM = "minLimit";
+    public static final String MAX_LIMIT_PARAM = "maxLimit";
+    public static final String REGION_PARAM = "region";
+    public static final String SORT_BY_PARAM = "sortBy";
+    public static final String ORDER_PARAM = "order";
+    public static final String PAGE_SIZE_PARAM = "pageSize";
+    public static final String PAGE_INDEX_PARAM = "pageIndex";
+
+    private JobsDatabase jobsDatabase;
+
+    @Override
+    public void init() {
+        this.jobsDatabase = new JobsDatabase();
+    }
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        try {
+            
+            JobQuery jobQuery = parseJobQuery(request);
+
+            JobPage jobPage = fetchJobPageDetails(jobQuery);
+
+            String json = new Gson().toJson(jobPage);
+            response.setContentType("application/json;");
+            response.getWriter().println(json);
+        } catch(IllegalArgumentException | ServletException | ExecutionException | TimeoutException e) {
+            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+        }
+    }
+
+    /**
+     * Returns the JobPage object.
+     *
+     * @param jobQuery The job query object with all the filtering/sorting params.
+     * @return JobPage object with all the details for the GET response.
+     */
+    private JobPage fetchJobPageDetails(JobQuery jobQuery) throws ServletException, ExecutionException, TimeoutException {
+        try {
+            return this.jobsDatabase.fetchJobPage(jobQuery)
+                    .get(TIMEOUT, TimeUnit.SECONDS);
+        } catch (InterruptedException | IOException e) {
+            throw new ServletException(e);
+        }
+    }
+
+    /**
+     * Returns the job query object.
+     *
+     * @param request From the GET request.
+     * @return the job query object.
+     * @throws IllegalArgumentException if one of the params is invalid.
+     */
+    public static JobQuery parseJobQuery(HttpServletRequest request) throws IllegalArgumentException {
```

Comment:
this tells me you are always sending all the filters up from the client and they all must have a value. If that's expected then great, otherwise your parseXXX(request) methods should return optional for min, max, region, etc. (and also you can update your JobQuery to also take on optional fields in setXXX so that you don't need ifs everywhere)

Reply:
yep the request will always have all filters with values so no need for optionals :)

Reply:
noted. thanks for clarifying.